### ISSUE TITLE
movie review

### ISSUE THREAD

**Issue by Baibhavi-rgh** [Baibhavi-rgh/CSE-Project#3]
Python

import pandas as pd
from scipy.sparse import csr_matrix
from sklearn.neighbors import NearestNeighbors

# --- 1. Load Data (Simplified Structure for Example) ---
# Dummy data simulating 'ratings.csv' (userId, movieId, rating)
ratings_data = {
    'userId': [1, 1, 1, 2, 2, 3, 3, 3, 4, 4],
    'movieId': [1, 2, 3, 1, 4, 2, 3, 5, 4, 5],
    'rating': [5.0, 4.0, 3.0, 5.0, 1.0, 4.0, 5.0, 2.0, 1.0, 5.0]
}
ratings_df = pd.DataFrame(ratings_data)

# Dummy data simulating 'movies.csv' for title lookup
movies_data = {
    'movieId': [1, 2, 3, 4, 5],
    'title': ['Toy Story (1995)', 'Jumanji (1995)', 'Waiting to Exhale (1995)', 'Four Rooms (1995)', 'Father of the Bride Part II (1995)'],
}
movies_df = pd.DataFrame(movies_data)

# --- 2. Create User-Item Matrix ---
# Pivot the ratings table to have movies as index and users as columns.
user_movie_matrix = ratings_df.pivot(
    index='movieId',
    columns='userId',
    values='rating'
).fillna(0) # Fill NaN with 0 (meaning no rating)

# Convert the DataFrame to a Sparse Matrix (more efficient for large datasets)
movie_features_matrix = csr_matrix(user_movie_matrix.values)

# --- 3. Build the KNN Model ---
# We use NearestNeighbors with 'cosine' distance to measure similarity between items (movies)
model_knn = NearestNeighbors(metric='cosine', algorithm='brute')
model_knn.fit(movie_features_matrix)

# --- 4. Recommendation Function ---
def get_collaborative_recommendations(movie_name, k=10):
    """Generates item-item collaborative recommendations using KNN."""
    if movie_name not in movies_df['title'].values:
        return f"Error: Movie '{movie_name}' not found."

    # Get the movieId
    movie_id = movies_df[movies_df['title'] == movie_name]['movieId'].iloc[0]

    # Get the index of the movie in the matrix
    movie_idx = user_movie_matrix.index.get_loc(movie_id)

    # Find the k nearest neighbors (similar movies)
    # distances are similarity, indices are the movie index in the matrix
    distances, indices = model_knn.kneighbors(
        user_movie_matrix.iloc[movie_idx, :].values.reshape(1, -1),
        n_neighbors=k + 1 # k neighbors + itself
    )

    # Extract the recommended movie titles and distances
    recommended_movies = []
    for i in range(1, len(distances.flatten())):
        # Get the actual movieId from the matrix index
        rec_movie_id = user_movie_matrix.index[indices.flatten()[i]]
        rec_movie_title = movies_df[movies_df['movieId'] == rec_movie_id]['title'].iloc[0]
        similarity = 1 - distances.flatten()[i] # 1 - distance = similarity
        recommended_movies.append((rec_movie_title, similarity))

    return pd.DataFrame(recommended_movies, columns=['Recommended Movie', 'Similarity Score'])

# --- 5. Test the System ---
print("\n---")
print("## Collaborative Filtering Recommendations for 'Toy Story (1995)'")
print(get_collaborative_recommendations('Toy Story (1995)', k=3))
